src/Capgemini.PowerApps.PackageDeployerTemplate/PackageTemplateBase.cs

```diff
@@ -1,148 +1,405 @@
         public override bool AfterPrimaryImport()
         {
-            this.PackageLog.Log($"{nameof(PackageTemplateBase)}.{nameof(this.AfterPrimaryImport)} running...", TraceEventType.Information);
-            if (this.ConfigDataStorage.ActivateDeactivateSLAs)
+            this.ExecuteLifecycleEvent(nameof(this.AfterPrimaryImport), () =>
             {
-                this.SlaDeploymentService.ActivateAll();
-                this.SlaDeploymentService.SetDefaultSlas(this.ConfigDataStorage.DefaultSlas);
-            }
+                if (this.ConfigDataStorage.ActivateDeactivateSLAs)
+                {
+                    this.SlaDeploymentService.ActivateAll();
+                    this.SlaDeploymentService.SetDefaultSlas(this.ConfigDataStorage.DefaultSlas);
+                }
 
-            this.ProcessDeploymentService.Deactivate(this.ConfigDataStorage.ProcessesToDeactivate);
-            this.SdkStepsDeploymentService.Deactivate(this.ConfigDataStorage.SdkStepsToDeactivate);
-            this.DataImporterService.Import(this.ConfigDataStorage.DataImports?.Where(c => !c.ImportBeforeSolutions), this.PackageFolderPath);
-            this.ProcessDeploymentService.Activate(this.ConfigDataStorage.ProcessesToActivate);
-            this.WordTemplateImporterService.ImportWordTemplates(this.ConfigDataStorage.WordTemplates, this.PackageFolderPath);
-            this.FlowActivationService.ActivateFlows(this.ConfigDataStorage.FlowsToDeactivate, this.ProcessedSolutions);
+                this.ProcessDeploymentService.Deactivate(this.ConfigDataStorage.ProcessesToDeactivate);
+                this.SdkStepDeploymentService.Deactivate(this.ConfigDataStorage.SdkStepsToDeactivate);
+                this.DataImporterService.Import(this.ConfigDataStorage.DataImports?.Where(c => !c.ImportBeforeSolutions), this.PackageFolderPath);
+                this.ProcessDeploymentService.Activate(this.ConfigDataStorage.ProcessesToActivate);
+                this.WordTemplateImporterService.ImportWordTemplates(this.ConfigDataStorage.WordTemplates, this.PackageFolderPath);
+                this.FlowDeploymentService.ConnectConnectionReferences(this.ConnectionReferenceMappings);
+                this.FlowDeploymentService.ActivateFlows(this.ConfigDataStorage.FlowsToDeactivate, this.ProcessedSolutions);
+            });
 
-            this.PackageLog.Log($"{nameof(PackageTemplateBase)}.{nameof(this.AfterPrimaryImport)} completed.", TraceEventType.Information);
             return true;
         }
+
+        /// <summary>
+        /// Gets a setting either from runtime arguments or an environment variable (in that order of preference). Environment variables should be prefixed with 'PACKAGEDEPLOYER_SETTINGS_'.
+        /// </summary>
+        /// <typeparam name="T">The type of argument.</typeparam>
+        /// <param name="key">The key.</param>
+        /// <returns>The setting value (if found).</returns>
+        protected T GetSetting<T>(string key)
```

Comment:
It could be. The only reason I kept it within the `PackageTemplateBase` is because `RuntimeSettings` is only exposed to sub-classes of `ImportExtension`. I'd need at minimum pass the `RuntimeSettings` to another class or expose it through a static property or something.

Reply:
Could it be another Service given the `RunSettings` and `TraceLogger`? Just an idea. @markcunninghamuk what do you think?

Reply:
I guess we could have something along the lines of `SettingsResolver`. It would probably making unit testing a bit easier.

Reply:
There probably isn't much scope for reusability while the `PackageTemplateBase` class is still responsible for translating these settings into the parameters that the services require. If we did the refactoring where `PackageTemplateBase` isn't responsible for all of that and instead just delegates to other classes that handle the lifecycle hooks themselves e.g. `FlowDeploymentExtension` then it would make sense to pass something like a `SettingsResolver`.

Reply:
That's reasonable - we can always refactor later.